src/java.base/share/classes/jdk/internal/util/DecimalDigits.java

```diff
@@ -132,6 +134,53 @@ public static int stringSize(long x) {
         return 19 + d;
     }
 
+    /**
+     * Determine whether the two strings in bytes are both numbers. If they are, return d0 * 10 + d1, otherwise return -1
+     * @param str The input LATIN1 encoded String value
+     * @param offset the offset
+     * @return If both characters are numbers, return d0 * 10 + d1, otherwise return -1
+     */
+    @ForceInline
+    public static int digit2(byte[] str, int offset) {
+        /*
+            Here we are doing a 2-Byte Vector operation on the short type.
+
+            x & 0xF0 != 0x30
+            ---------------
+            0 0b0011_0000 & 0b1111_0000 = 0b0011_0000
+            1 0b0011_0001 & 0b1111_0000 = 0b0011_0000
+            2 0b0011_0010 & 0b1111_0000 = 0b0011_0000
+            3 0b0011_0011 & 0b1111_0000 = 0b0011_0000
+            4 0b0011_0100 & 0b1111_0000 = 0b0011_0000
+            5 0b0011_0101 & 0b1111_0000 = 0b0011_0000
+            6 0b0011_0110 & 0b1111_0000 = 0b0011_0000
+            7 0b0011_0111 & 0b1111_0000 = 0b0011_0000
+            8 0b0011_1000 & 0b1111_0000 = 0b0011_0000
+            9 0b0011_1001 & 0b1111_0000 = 0b0011_0000
+
+            (((d = x & 0x0F) + 0x06) & 0xF0) != 0
+            ---------------
+            0 ((0b0011_0000) & 0b0000_1111 + 0b0110_0000) & 0b1111_0000 = 0b0110_0000
+            1 ((0b0011_0001) & 0b0000_1111 + 0b0110_0000) & 0b1111_0000 = 0b0110_0000
+            2 ((0b0011_0010) & 0b0000_1111 + 0b0110_0000) & 0b1111_0000 = 0b0110_0000
+            3 ((0b0011_0011) & 0b0000_1111 + 0b0110_0000) & 0b1111_0000 = 0b0110_0000
+            4 ((0b0011_0100) & 0b0000_1111 + 0b0110_0000) & 0b1111_0000 = 0b0110_0000
+            5 ((0b0011_0101) & 0b0000_1111 + 0b0110_0000) & 0b1111_0000 = 0b0110_0000
+            6 ((0b0011_0110) & 0b0000_1111 + 0b0110_0000) & 0b1111_0000 = 0b0110_0000
+            7 ((0b0011_0111) & 0b0000_1111 + 0b0110_0000) & 0b1111_0000 = 0b0110_0000
+            8 ((0b0011_1000) & 0b0000_1111 + 0b0110_0000) & 0b1111_0000 = 0b0110_0000
+            9 ((0b0011_1001) & 0b0000_1111 + 0b0110_0000) & 0b1111_0000 = 0b0110_0000
+         */
+        int d;
+        short x = UNSAFE.getShortUnaligned(str, Unsafe.ARRAY_BYTE_BASE_OFFSET + offset, false);
```

Comment:
@wenshao I'm a bit worried about the use of Unsafe here.

This method is `public` (although in an internal package), and while it is used correctly in this PR, there's no warning in the doc that `str` and `offset` must come from a trusted caller that ensures that they are safe to use with, well..., Unsafe.

Did you consider safer alternatives, like usage of `VarHandle`, even if that might mean a slight performance degradation?

Reply:
Never mind, `VarHandle` cannot be used in this case.

Reply:
Using VarHandle in this scenario may affect JVM startup performance

Reply:
I concur that unsafe usage here is undesirable. Performance is not the only metric of interest when working on the JDK core libraries.